### PR TITLE
Generalize generation infrastructure to account for mixed types (FP8)

### PR DESCRIPTION
**Pull request by krzysz00**
1. In the ConvGenerator, expand on the work from #1087 to have proper filter, input, and output data type strings.
2. Amend rocmlir-gen to take --fil_dtype/--a_dtype/-tf/-ta, --in_dtype/-b_dtype/-ti/-tb, and --out_dtype/--c_dtype/-to/-tc to allow for specifying any combination of types for a convolution or gemm.
2b. Make the `-t` option an alias for quickly setting `-t{a,b,c}`, including the 8-bit {int,float} input  -> 32-bit {int,float} output default.
3. While doing the above, extend the various strToType() methods to accept 8-bit float types.
4. Make the GenParams struct in rocmilr-gen take a vector of types to account for our new generality.
5. Instead of hard-coding f16 and bf16 as the small floats that use f32 for validation, send all floats of size < 32 to f32. That way we can't miss a case.

### CODE REVIEW

**Comment by krzysz00** [comment]
Note that, after this PR lands, we'll be able to run an fp8 kernel on MI-300, I think.
With #1079 and #1088, we'll be close to being able to do a GPU validation (probably).
We may end up needing to write CPU utility functions analogous to the ones used for f16 in order to get everything to codegen.

**Comment by yiqian1** [comment]
LGTM

**Comment by krzysz00** [comment]
@yiqian1 I think you forgot to hit approve?

**Review by yiqian1 — APPROVED** [review]

File: mlir/lib/Dialect/Rock/Generator/Conv2dGenerator.cpp
Comment (jerryyin) [inline mlir/lib/Dialect/Rock/Generator/Conv2dGenerator.cpp:708]: This comment doesn't hold off merging.

I think you are just extending the original logic here. But this will tentatively set the data type to (potentially) wrong types, and we leave those field as-is. Then later, when the "real" data type is required, we do a second computation regardless of what the field value is using `getFilterDataType()`, `getInputDataType()` and `getOutputDataType()`.

I would much more prefer instead
 - setDataTypes() does everything, and populate the correct types for filter/intput/output member fields at one go
 - get*Type() only return the field and doesn't do any redundant derive process.

Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Generator/Conv2dGenerator.cpp:708]: This API is literally for `setDataTypes("f32");` for verification code.

**Pull request merged**
